RadClass/H0.py

```diff
@@ -0,0 +1,121 @@
+        self.gross = gross
+        self.x1 = None
+
+        # arrays are structured as matrices (nxm) for writing to file
+        if self.gross:
+            # four values are saved: timestamp, x1, x2, and p-val
+            self.triggers = np.empty((0, 4))
+        else:
+            # all p-vals for rejected hypothesis are saved in this sparse array
+            self.triggers = np.empty((0, energy_bins+1))
+
+    def run_gross(self, data, timestamp):
+        data = np.sum(data)
+
+        # only needed for the first initialization
+        if self.x1 is None:
+            self.x1 = data
+        else:
+            self.x2 = data
+            n = self.x1 + self.x2
+            p = 0.5
+            pval = stats.binom_test(self.x2, n, p, alternative='two-sided')
+
+            # only save instances with rejected null hypothesesf
+            if pval <= self.significance:
+                self.triggers = np.append(self.triggers,
+                                            [[timestamp, pval,
+                                            self.x1, self.x2]],
+                                            axis=0)
+
+            # saving data for the next integration step
+            self.x1 = self.x2
+    
+    def run_channels(self, data, timestamp):
+        # only needed for the first initialization
+        if self.x1 is None:
+            self.x1 = data
+        else:
+            self.x2 = data
+            rejections = np.ones_like(data)
+            nvec = self.x1 + self.x2
+            p = 0.5
+            for i, (x1, n) in enumerate(zip(self.x1, nvec)):
+                pval = stats.binom_test(x1, n, p,
+                                            alternative='two-sided')
+                if pval <= self.significance:
+                    rejections[i] = pval
+            #if np.nonzero(rejections)[0].size != 0:
+            if np.sum(rejections) != len(rejections):
+                self.triggers = np.append(self.triggers,
+                                        [np.insert(rejections,
+                                                    0, timestamp)],
+                                        axis=0)
+
+            # saving data for the next integration step
+            self.x1 = self.x2
+
+    run_method = { True: run_gross, False: run_channels }
+
+    def run(self, data, timestamp):
+        '''
+        Method required by RadClass. Called at each integration step.
+        Completes hypothesis testing for passed data and stores results.
+        Wrapper method that chooses run_gross or run_channel based on user
+        input variable: gross.
+        '''
+        self.run_method[self.gross](self, data, timestamp)
+
+    def write(self, filename):
+        '''
+        Writes results of hypothesis test to file by constructing a Pandas
+        DataFrame and saving to csv with name filename.
+        '''
+
+        results = pd.DataFrame()
+        results['timestamps'] = self.triggers[:, 0]
+        if len(self.triggers[0]) == 4:
```

Comment:
I think it would be overkill, since only a handful of lines are required to construct the DataFrame and write to file. If this was a larger method, I would certainly agree with separating it into different methods.